### DISCUSSION TITLE
Cannot load data to sql database

### DISCUSSION THREAD

**Discussion by Edimo05** [mkleehammer/pyodbc#1159] [Q&A] [answered]
Please first make sure you have looked at:

* Documentation: https://github.com/mkleehammer/pyodbc/wiki
* Other issues

### Environment

To diagnose, we usually need to know the following, including version numbers.  On Windows, be
sure to specify 32-bit Python or 64-bit:

- Python: python 3.10.6 64-bit
- pyodbc: 4.0.34
- OS: Windows
- DB: SQL Server
- driver: ODBC Driver 17 for SQL Server

ERROR: (pyodbc.OperationalError) ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Named Pipes Provider: Could not open a connection to SQL Server [53].  (53) (SQLDriverConnect); [08001] [Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired (0); [08001] [Microsoft][ODBC Driver 17 for SQL Server]A network-related or instance-specific error has occurred while establishing a connection to SQL Server. Server is not found or not accessible. Check if instance name is correct and if SQL Server is configured to allow remote connections. For more information see SQL Server Books Online. (53)')

Code:
user = 'Periquito'
password = 'PinPinPinPan'
server = 'internal-tools-db.devfacebook.com'
dbname = 'AdManagementTool'
driver='{ODBC Driver 17 for SQL Server}'

df.to_sql(table, f"mssql+pyodbc://ERRSTSDBP2/{user}:{password}@{server}/{dbname}?driver=ODBC+Driver+17+for+SQL+Server", if_exists='append', index=False)

I already stablished a connection with the server in other codes, using pyodbc, not sure why is not working just in this one.

**Comment by v-chojas** [comment]
It doesn't look like you're using pyODBC directly, could you try connecting directly?

**Comment by Edimo05** [comment]
@gordthompson , I got the same error:  Expected URI to be a string, got <class 'sqlalchemy.engine.base.Engine'>.
@v-chojas if I connect directly, the same error as above appears.

**Comment by gordthompson** [comment]
Leave pandas' `.to_sql()` out of the mix for now. Does this code work?

```python
import pyodbc

user = 'Periquito'
password = 'PinPinPinPan'
server = 'internal-tools-db.devfacebook.com'
dbname = 'AdManagementTool'
driver='{ODBC Driver 17 for SQL Server}'

connection_string = (
    f"Driver={driver};"
    f"Server={server};"
    f"Database={dbname};"
    f"UID={user};"
    f"PWD={password};"
)
cnxn = pyodbc.connect(connection_string)
crsr = cnxn.cursor()
result = crsr.execute("SELECT 'Connected!' AS foo").fetchval()
print(result)
```

If not, then please post the complete stack trace.

**Reply by Edimo05** [reply]
hi, message: Connected!
![image](https://user-images.githubusercontent.com/94873872/215802597-c3912905-8ef0-4317-a18a-7f6f92816a5f.png)

**Reply by gordthompson** [reply]
So this is not a pyodbc issue. It is an issue with SQLAlchemy, or pandas, or something else.

**Comment by gordthompson** [comment, marked as answer]
This is not a pyodbc issue. It is an issue with SQLAlchemy, or pandas, or something else.
